ports/alif/machine_pin.c

```diff
@@ -225,6 +288,125 @@ static mp_obj_t machine_pin_toggle(mp_obj_t self_in) {
+        irq->base.methods = (mp_irq_methods_t *)&machine_pin_irq_methods;
+        irq->base.parent = MP_OBJ_FROM_PTR(self);
+        irq->base.handler = mp_const_none;
+        irq->base.ishard = false;
+        irq->irq_num = (self->port < 15) ? (GPIO0_IRQ0_IRQn + irq_idx) : (LPGPIO_IRQ0_IRQn + self->pin);
+        irq->reserved = false;
+        MP_STATE_PORT(machine_pin_irq_obj[irq_idx]) = irq;
+    }
+
+    return irq;
+}
+
+static mp_uint_t machine_pin_irq_trigger(mp_obj_t self_in, mp_uint_t trigger) {
+    machine_pin_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    machine_pin_irq_obj_t *irq = machine_pin_get_irq(self);
+
+    irq->flags = 0;
+    irq->trigger = trigger;
+
+    if (trigger == 0) {
+        // Disable IRQs.
+        gpio_disable_interrupt(self->gpio, self->pin);
+        gpio_mask_interrupt(self->gpio, self->pin);
+
+        NVIC_ClearPendingIRQ(irq->irq_num);
+        NVIC_DisableIRQ(irq->irq_num);
+        return 0;
+    }
+
+    // Clear and enable GPIO IRQ.
+    gpio_enable_interrupt(self->gpio, self->pin);
+    gpio_unmask_interrupt(self->gpio, self->pin);
+
+    // Configure GPIO IRQ trigger
+    if (trigger == MP_HAL_PIN_TRIGGER_FALL) {
+        gpio_interrupt_set_edge_trigger(self->gpio, self->pin);
+        gpio_interrupt_set_polarity_low(self->gpio, self->pin);
+    } else if (trigger == MP_HAL_PIN_TRIGGER_RISE) {
+        gpio_interrupt_set_edge_trigger(self->gpio, self->pin);
+        gpio_interrupt_set_polarity_high(self->gpio, self->pin);
+    } else if (trigger == (MP_HAL_PIN_TRIGGER_FALL | MP_HAL_PIN_TRIGGER_RISE)) {
+        gpio_interrupt_set_both_edge_trigger(self->gpio, self->pin);
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("Invalid IRQ trigger"));
+    }
+
+    // Clear GPIO IRQ (must be done after configuring trigger)
+    gpio_interrupt_eoi(self->gpio, self->pin);
+
+    // Clear and enable NVIC GPIO IRQ.
+    NVIC_ClearPendingIRQ(irq->irq_num);
+    NVIC_SetPriority(irq->irq_num, IRQ_PRI_GPIO);
+    NVIC_EnableIRQ(irq->irq_num);
+    return 0;
+
+}
+
+static mp_uint_t machine_pin_irq_info(mp_obj_t self_in, mp_uint_t info_type) {
+    machine_pin_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    machine_pin_irq_obj_t *irq = machine_pin_get_irq(self);
+
+    if (info_type == MP_IRQ_INFO_FLAGS) {
+        return irq->flags;
+    } else if (info_type == MP_IRQ_INFO_TRIGGERS) {
+        return irq->trigger;
+    }
+    return 0;
+}
+
+static const mp_irq_methods_t machine_pin_irq_methods = {
+    .trigger = machine_pin_irq_trigger,
+    .info = machine_pin_irq_info,
+};
+
+// pin.irq(handler=None, trigger=IRQ_FALLING|IRQ_RISING, hard=False)
+static mp_obj_t machine_pin_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_handler, ARG_trigger, ARG_hard };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_handler, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
+        { MP_QSTR_trigger, MP_ARG_INT, {.u_int = MP_HAL_PIN_TRIGGER_FALL | MP_HAL_PIN_TRIGGER_RISE} },
```

Comment:
I tried setting the trigger to rising only, and falling only, but in both those cases it seems to always trigger on both edges.

Can you test to see if you can get it triggering on just one edge?

Reply:
Yes, it works really well if you enable:
```
gpio_enable_debounce(self->gpio, self->pin);
```